searcher.go

```diff
@@ -0,0 +1,168 @@
+package certificates
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// watchTimeOut is the time to wait on watches against the Kubernetes API
+	// before giving up and throwing an error.
+	watchTimeOut = 90 * time.Second
+)
+
+type Config struct {
+	// Dependencies.
+
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		K8sClient: nil,
+		Logger:    nil,
+	}
+}
+
+func NewSearcher(config Config) (*Searcher, error) {
+	// Dependencies.
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+
+	s := &Searcher{
+		// Dependencies.
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return s, nil
+}
+
+type Searcher struct {
+	// Dependencies.
+
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+}
+
+func (s *Searcher) SearchCluster(clusterID string) (Cluster, error) {
+	var cluster Cluster
+	var err error
+
+	err = s.searchError(&cluster.APIServer, clusterID, apiCert, err)
+	err = s.searchError(&cluster.CalicoClient, clusterID, calicoCert, err)
+	err = s.searchError(&cluster.EtcdServer, clusterID, etcdCert, err)
+	err = s.searchError(&cluster.ServiceAccount, clusterID, serviceAccountCert, err)
+	err = s.searchError(&cluster.Worker, clusterID, workerCert, err)
```

Comment:
IMO there should be the usual error handling or a list of assets we iterate over. Then we would only type the error handling ones. I do not like calling functions with errors to just make then return the error. That is unnecessary. We can do better.

Reply:
I'll try with loop.

Reply:
The loop is there.